packages/functional-tests/lib/metrics.ts

```diff
@@ -0,0 +1,127 @@
+// uid transformed downstream to user_id
+const P1_SUBSCRIPTION_PLAN_CHANGE_REQUIRED_FIELDS = [
+  'paymentProvider',
+  'previousPlanId',
+  'previousProductId',
+  'subscriptionId',
+  'uid',
+];
+
+const REQUIRED_FIELDS_BY_EVENT_TYPE_MAP = {
+  // subscription create
+  'amplitude.subPaySetup.view': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_CREATE_REQUIRED_FIELDS,
+  ],
+  'amplitude.subPaySetup.engage': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_CREATE_REQUIRED_FIELDS,
+  ],
+  'amplitude.subPaySetup.submit': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_CREATE_REQUIRED_FIELDS,
+    'paymentProvider',
+  ],
+  'amplitude.subPaySetup.success': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_CREATE_REQUIRED_FIELDS,
+    'paymentProvider',
+    'country_code_source',
+  ],
+  'amplitude.subPaySetup.fail': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_CREATE_REQUIRED_FIELDS,
+    'paymentProvider',
+    'error_id',
+  ],
+  // subscription plan change
+  'amplitude.subPaySubChange.view': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_PLAN_CHANGE_REQUIRED_FIELDS,
+  ],
+  'amplitude.subPaySubChange.engage': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_PLAN_CHANGE_REQUIRED_FIELDS,
+  ],
+  'amplitude.subPaySubChange.submit': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_PLAN_CHANGE_REQUIRED_FIELDS,
+  ],
+  'amplitude.subPaySubChange.success': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_PLAN_CHANGE_REQUIRED_FIELDS,
+  ],
+  'amplitude.subPaySubChange.fail': [
+    ...P1_PAYMENTS_GLOBAL_REQUIRED_FIELDS,
+    ...P1_SUBSCRIPTION_PLAN_CHANGE_REQUIRED_FIELDS,
+    'error_id',
+  ],
+};
+
+export class MetricsObserver {
+  public rawEvents: (Object & { type: string })[];
+  private subscribePage: SubscribePage;
+
+  constructor(subscribePage: SubscribePage) {
+    this.subscribePage = subscribePage;
+    this.rawEvents = [];
+  }
+
+  startTracking() {
+    this.subscribePage.page.on('request', (request) => {
+      if (request.method() === 'POST' && request.url().includes('/metrics')) {
+        const requestBody = request.postDataJSON();
+
+        for (const eventBase of requestBody.events) {
+          const eventProperties = structuredClone(requestBody.data);
+          const rawEvent = { ...eventBase, ...eventProperties };
+
+          if (rawEvent.type in REQUIRED_FIELDS_BY_EVENT_TYPE_MAP) {
+            for (const field of REQUIRED_FIELDS_BY_EVENT_TYPE_MAP[
+              rawEvent.type
+            ]) {
+              expect(rawEvent[field]).not.toBeNull();
+              expect(rawEvent[field], `${field} is required`).toBeDefined();
+            }
+
+            // If this is an existing user checkout flow or a plan change event, uid is required
+            if (
+              rawEvent.checkoutType === 'with-account' ||
```

Comment:
Non blocking -- I just remembered Ivo created an enum for this -- I think in fxa-shared. Ideally we could use that instead of hardcoding the value -- of course we're hardcoding a bunch of properties elsewhere at the moment.

Maybe we can include this in FXA-7364.